app/controllers/passwords_controller.rb

```diff
@@ -78,22 +65,44 @@ def ensure_email_present
   end
 
   def validate_confirmation_token
-    @user = User.find_by(confirmation_token: params[:token].to_s)
-    redirect_to root_path, alert: t("passwords.edit.token_failure") unless @user&.valid_confirmation_token?
+    confirmation_token = params.permit(:token).fetch(:token, "").to_s
+    @user = User.find_by(confirmation_token:)
+    return login_failure(t("passwords.edit.token_failure")) unless @user&.valid_confirmation_token?
+    sign_out if signed_in? && @user != current_user
```

Comment:
should this be in an after_action filter? i.e. is it applicable to other actions as well

Reply:
the after_action doesn't always run when it needs to.

I first tried removing this code entirely and saw that it didn't make any tests fail, so I wrote a test for each of the possible branches here, made sure they fail without the code, and then tried the refactor you suggested and it doesn't sign out the user correctly. I'll leave it here and I'm glad you made this suggestion since it found a test coverage gap.